download/upgradeRecipe/upgradeRecipe8.adoc

```diff
@@ -228,3 +273,722 @@ After in `*.xml`:
+The logic of this new filter is therefore inverted as compared to the old filter.
+
+Users should update their `@PlanningEntity` annotations, supplying the new filter instead of the old.
+The old field has now been removed.
+
+Before in `*.java`:
+
+[source,java]
+----
+@PlanningEntity(movableEntitySelectionFilter = MyMovableEntitySelectionFilter.class)
+----
+
+After in `*.java`:
+
+[source,java]
+----
+@PlanningEntity(pinningFilter = MyPinningFilter.class)
+----
+
+[.upgrade-recipe-minor.upgrade-recipe-public-api]
+=== `@PlanningVariable.reinitializeVariableEntityFilter` removed
+
+The `reinitializeVariableEntityFilter` field on `@PlanningVariable` annotation has been deprecated for removal in 7.x and now removed.
+
+Users of this niche functionality should refer to the documentation on how to achieve the same result by
+link:https://docs.optaplanner.org/latest/optaplanner-docs/html_single/index.html#nullablePlanningVariable[power-tweaking construction heuristics].
+
+[.upgrade-recipe-minor.upgrade-recipe-public-api]
+=== `*ScoreHolder` classes turned into interfaces
+
+In OptaPlanner 7, `ScoreHolder` classes, used exclusively for
+https://docs.optaplanner.org/latest/optaplanner-docs/html_single/index.html#droolsScoreCalculation[Drools score calculation], exposed a number of public methods which, if used, allowed the user to unintentionally corrupt or otherwise negatively affect their scores.
+
+In OptaPlanner 8, these methods have been removed and the classes have been turned into interfaces.
+For the vast majority of users who did not use any of these potentially harmful methods, there will be no impact on their code.
+Others will find suitable replacements in the public API in areas of
+https://docs.optaplanner.org/latest/optaplanner-docs/html_single/index.html#explainingTheScore[score explanation] and
+https://docs.optaplanner.org/latest/optaplanner-docs/html_single/index.html#constraintConfiguration[constraint configuration].
+
+[.upgrade-recipe-minor]
+=== `ValueRangeFactory` class now final
+
+`ValueRangeFactory` class is a factory class that has only static methods.
+As such, there is no need for the users to extend this class, and it has therefore been made `final`.
+
+Before in `*.java`:
+
+[source,java]
+----
+class MyValueRangeFactory extends ValueRangeFactory {
+    ...
+}
+----
+
+After in `*.java`:
+
+[source,java]
+----
+class MyValueRangeFactory {
+    ...
+}
+----
+
+[.upgrade-recipe-minor.upgrade-recipe-public-api]
+=== `ConstraintMatchTotal` and `Indictment` now interfaces and immutable
+
+`ConstraintMatchTotal` and `Indictment` classes have been converted into interfaces and in the process, their implementations were moved out of the public API, together with methods that allowed to mutate their state.
+
+These methods were never intended for public API.
+Users who created their own instances of these classes, or who mutated state of the existing ones in any way, have likely experienced various unrecoverable issues.
```

Comment:
In ConstraintMatchAwareIncrementalScoreCalculator we require users to create their own.
We should have a before/after for that, using the DefaultConstraintMatch etc.

Reply:
Note that ConstraintMatchAwareIncrementalScoreCalculator isn't commonly used enough to start requiring a way to construct ConstraintMatch from the public API.